lightgbm_ranker.py

```diff
@@ -0,0 +1,192 @@
+        if self.categorical_match_features:
+            match_feature_names += self.categorical_match_features
+        for doc in docs:
+            query_feature = []
+            match_feature = []
+            query_values = [doc.tags.get(feature) for feature in query_feature_names]
+            for match in doc.matches:
+                match_values = [
+                    match.tags.get(feature) for feature in match_feature_names
+                ]
+                labels.append(match.tags.get(self.label))
+                match_feature.append(match_values)
+                query_feature.append(query_values)
+            group.append(len(doc.matches))
+            q_features.append(query_feature)
+            m_features.append(match_feature)
+
+        query_dataset = lightgbm.Dataset(
+            data=np.vstack(q_features),
+            group=group,
+            feature_name=query_feature_names,
+            categorical_feature=self.categorical_query_features,
+            free_raw_data=False,
+        )
+
+        match_dataset = lightgbm.Dataset(
+            data=np.vstack(m_features),
+            group=group,
+            label=labels,
+            feature_name=match_feature_names,
+            categorical_feature=self.categorical_match_features,
+            free_raw_data=False,
+        )
+        if task == 'predict':
+            return query_dataset.construct().add_features_from(
+                match_dataset.construct()
+            )
+        else:
+            return match_dataset.construct().add_features_from(
+                query_dataset.construct()
+            )
+
+    @requests(on='/train')
+    def train(self, docs: DocumentArray, **kwargs):
+        """The :meth:`train` endpoint allows user to train the lightgbm ranker
+        in an incremental manner. The features will be extracted from the `attr`:`tags`,
+        including all the :attr:`query_features` and :attr:`match_features`. The label/groundtruth of the
+        training data will be the :attr:`label` field.
+
+        :param docs: :class:`DocumentArray` passed by the user or previous executor.
+        :param kwargs: Additional key value arguments.
+        """
+        train_set = self._get_features_dataset(docs, task='train')
+        categorical_feature = []
+        if self.categorical_query_features:
+            categorical_feature += self.categorical_query_features
+        if self.categorical_match_features:
+            categorical_feature += self.categorical_match_features
+        if not categorical_feature:
+            categorical_feature = 'auto'
+        self.booster = lightgbm.train(
+            train_set=train_set,
+            init_model=self.booster,
+            params=self.params,
+            keep_training_booster=True,
+            categorical_feature=categorical_feature,
+        )
+
+    @requests(on='/dump')
+    def dump(self, **kwargs):
+        self.booster.save_model(self.model_path)
```

Comment:
what if `self.model_path` is None?

According to ur code this can be None, other thing is if u do `self.model_path = model_path or  'tmp/model.txt'